neuraxle/base.py

```diff
@@ -29,14 +29,55 @@
 from neuraxle.hyperparams.space import HyperparameterSpace, HyperparameterSamples
 
 
-class BaseStep(ABC):
+class Hasher(ABC):
+    def hash(self, hyperparameters: HyperparameterSamples, data_inputs: Any):
+        return hash(data_inputs)
+
+    @abstractmethod
+    def rehash(self, ids, hyperparameters: HyperparameterSamples, data_inputs: Any):
+        raise NotImplementedError()
+
+
+class HasherByIndex(Hasher):
+    def hash(self, hyperparameters, data_inputs: Any):
+        return range(len(data_inputs))
+
+    def rehash(self, ids, hyperparameters: HyperparameterSamples, data_inputs: Any):
+        return self.hash(hyperparameters, data_inputs)
+
+
+class IDHasher(Hasher):
+    def rehash(self, ids, hyperparameters: HyperparameterSamples, data_inputs: Any):
+        return ids
 
+
+class DataContainer:
+    def __init__(self, ids, data_inputs: Any, expected_outputs: Any = None):
+        self.original_ids = ids
+        self.ids = ids
+        self.data_inputs = data_inputs
+        self.expected_outputs = expected_outputs
+
+    def set_data_inputs(self, data_inputs: Any):
+        self.data_inputs = data_inputs
+
+    def set_ids(self, ids: Any):
+        self.ids = ids
+
+
+class BaseStep(ABC):
     def __init__(
             self,
             hyperparams: HyperparameterSamples = None,
             hyperparams_space: HyperparameterSpace = None,
-            name: str = None
+            name: str = None,
+            hasher: Hasher = None
```

Comment:
do we want the IDHasher by defaut and not permit None ? to avoid having to raise the exceptions below (see raise_hyperparams_without_hasher_exception())

Reply:
as discussed, only take the HasherByHyperparam